src/Common/serde.h

```diff
@@ -0,0 +1,49 @@
+#pragma once
+
+#include <IO/VarInt.h>
+#include <base/types.h>
+#include <Common/VersionRevision.h>
+
+namespace DB
+{
+/// REQUIRES: The object must support versioned serialization/deserialization
+template <typename S, typename WB>
+concept VersionedSerializable = requires(const S & s, WB && wb, VersionType version) { s.serialize(wb, version); };
+
+template <typename S, typename RB>
+concept VersionedDeserializable = requires(S && s, RB && rb, VersionType version) { s.deserialize(rb, version); };
+
+template <typename T>
+concept HasGetVersionFromRevision = requires(T && a) {
+    {
+        a.getVersionFromRevision(ProtonRevision::getVersionRevision())
+    } -> std::convertible_to<VersionType>;
+};
+
+template <typename WB, VersionedSerializable<WB> S>
+void ALWAYS_INLINE serialize(const S & s, WB && wb)
+{
+    /// By default use `Revision` for version management, unless it has its own versions
+    if constexpr (HasGetVersionFromRevision<S>)
+        s.serialize(wb, static_cast<VersionType>(s.getVersionFromRevision(ProtonRevision::getVersionRevision())));
+    else
+        s.serialize(wb, static_cast<VersionType>(ProtonRevision::getVersionRevision()));
+}
+
+template <typename RB, VersionedDeserializable<RB> S>
+void ALWAYS_INLINE deserialize(S & s, RB && rb)
```

Comment:
It seems we have a `&&` and we reading data from  it.

I still we have too much complexity here

Reply:
1) change `&&` to `&`, it's make sense

2) I think it's no slowness, and no confusion, which just used to unify the serialization/deserialization interface to simplify the interface call of version management objects

Reply:
It's a little complicated, I'll readjust it